File: motor3d 2.0/quadtree.vb

Namespace Espacio2D
    Public Class Quadtree
        Inherits ObjetoGeometrico2D

        Private mNiveles As Integer

        Private mSectorRaiz As SectorQuadtree

        Public Property Espacio() As AABB2D
            Get
                Return mSectorRaiz.Espacio
            End Get
            Set(ByVal value As AABB2D)
                If mSectorRaiz.Espacio <> value Then
                    mSectorRaiz = New SectorQuadtree(mNiveles, value)
                End If
            End Set
        End Property

        Public Property Niveles() As Integer
            Get
                Return mNiveles
            End Get
            Set(ByVal value As Integer)
                If value > 0 Then
                    mNiveles = value
                    mSectorRaiz = New SectorQuadtree(mNiveles, mSectorRaiz.Espacio)
                End If
            End Set
        End Property

        Public ReadOnly Property SectorRaiz() As SectorQuadtree
            Get
                Return mSectorRaiz
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Punto As Punto2D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Pertenece(Punto)
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal AABB As AABB2D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Colision(AABB)
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Figura2D As Figura2D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Colision(Figura2D.AABB)
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Recta As Recta2D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Pertenece(Recta)
            End Get
        End Property

        Public Sub New(ByVal Niveles As Integer, ByVal AABB As AABB2D)
            If Niveles > 1 Then
                mNiveles = Niveles

                mSectorRaiz = New SectorQuadtree(Niveles, AABB)
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (NEW): Un octree debe tener al menos dos niveles:" & vbNewLine _
                                                & "Niveles=" & Niveles)
            End If
        End Sub

        Public Sub Refrescar()
            mSectorRaiz.Refrescar()
        End Sub

        Public Function Sector(ByVal Punto As Punto2D) As SectorQuadtree
            Return Sector(Me, Punto)
        End Function

        Public Function Sector(ByVal Recta As Recta2D) As SectorQuadtree
            Return Sector(Me, Recta)
        End Function

        Public Function Sector(ByVal AABB As AABB2D) As SectorQuadtree
            Return Sector(Me, AABB)
        End Function

        Public Function Sector(ByVal Figura2D As Figura2D) As SectorQuadtree
            Return Sector(Me, Figura2D)
        End Function

        Public Function Sectores(ByVal Recta As Recta2D) As SectorQuadtree()
            Return Sectores(Me, Recta)
        End Function

        Public Shared Function Sector(ByVal Quadtree As Quadtree, ByVal Punto As Punto2D) As SectorQuadtree
            Dim Resultado As Integer = Quadtree.SectorRaiz.Pertenece(Punto)
            Dim S As SectorQuadtree = Quadtree.SectorRaiz

            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return S
                Else
                    Do While Resultado <> -2 AndAlso Resultado <> -1
                        Resultado = S.Pertenece(Punto)
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.SubSectores(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                    Loop

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (PERTENECE): El punto especificado no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Punto=" & Punto.ToString & vbNewLine _
                                                & "Espacio=" & Quadtree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sector(ByVal Quadtree As Quadtree, ByVal Recta As Recta2D) As SectorQuadtree
            Dim Resultado As Integer
            Dim S As SectorQuadtree = Quadtree.SectorRaiz

            Resultado = Quadtree.SectorRaiz.Pertenece(Recta)

            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return S
                Else
                    Do While Resultado <> -2 AndAlso Resultado <> -1
                        Resultado = S.Pertenece(Recta)
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.SubSectores(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                    Loop

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (PERTENECE): La recta especificada no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Punto=" & Recta.ToString & vbNewLine _
                                                & "Espacio=" & Quadtree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sectores(ByVal Quadtree As Quadtree, ByVal Recta As Recta2D) As SectorQuadtree()
            Dim S As SectorQuadtree = Quadtree.SectorRaiz
            Dim Retorno As New List(Of SectorQuadtree)

            If Quadtree.SectorRaiz.Espacio.Pertenece(Recta) Then
                InterseccionRecta(Recta, S, Retorno)

                Return Retorno.ToArray
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (PERTENECE): La recta especificado no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Recta=" & Recta.ToString & vbNewLine _
                                                & "Espacio=" & Quadtree.Espacio.ToString)
            End If
        End Function

        Public Shared Sub InterseccionRecta(ByVal Recta As Recta2D, ByVal Sector As SectorQuadtree, ByRef ListaRetorno As List(Of SectorQuadtree))
            If Not Sector.EsHoja Then
                For Each Hijo As SectorQuadtree In Sector.SubSectores
                    If Hijo.Espacio.Pertenece(Recta) Then
                        InterseccionRecta(Recta, Hijo, ListaRetorno)
                    End If
                Next
            Else
                If Sector.Espacio.Pertenece(Recta) Then ListaRetorno.Add(Sector)
            End If
        End Sub

        Public Shared Function Sector(ByVal Quadtree As Quadtree, ByVal AABB As AABB2D) As SectorQuadtree
            Dim Resultado As Integer = Quadtree.SectorRaiz.Pertenece(AABB)
            Dim S As SectorQuadtree = Quadtree.SectorRaiz

            'PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return Quadtree.SectorRaiz
                Else
                    Do
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.SubSectores(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                        Resultado = S.Pertenece(AABB)
                    Loop While Resultado <> -2 AndAlso Resultado <> -1

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (PERTENECE): La AABB especificada no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "AABB=" & AABB.ToString & vbNewLine _
                                                & "Espacio=" & Quadtree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sector(ByVal Quadtree As Quadtree, ByRef Figura2D As Figura2D) As SectorQuadtree
            Dim Resultado As Integer = Quadtree.SectorRaiz.Pertenece(Figura2D)
            Dim S As SectorQuadtree = Quadtree.SectorRaiz

            'PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return Quadtree.SectorRaiz
                Else
                    Do
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.SubSectores(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                        Resultado = S.Pertenece(Figura2D)
                    Loop While Resultado <> -2 AndAlso Resultado <> -1

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica2D("OCTREE (PERTENECE): La figura especificada no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Figura2D=" & Figura2D.ToString & vbNewLine _
                                                & "Espacio=" & Quadtree.Espacio.ToString)
            End If
        End Function

        Public Overrides Function ToString() As String
            Return "{Quadtree de espacio=" & mSectorRaiz.Espacio.ToString & "}"
        End Function
    End Class
End Namespace

File: motor3d 2.0.ConvertedToC#/Quadtree.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;

namespace Motor3D.Espacio2D
{
	public class Quadtree : ObjetoGeometrico2D
	{


		private int mNiveles;

		private SectorQuadtree mSectorRaiz;
		public Caja2D Espacio {
			get { return mSectorRaiz.Espacio; }
			set {
				if (mSectorRaiz.Espacio != value) {
					mSectorRaiz = new SectorQuadtree(mNiveles, value);
				}
			}
		}

		public int Niveles {
			get { return mNiveles; }
			set {
				if (value > 0) {
					mNiveles = value;
					mSectorRaiz = new SectorQuadtree(mNiveles, mSectorRaiz.Espacio);
				}
			}
		}

		public SectorQuadtree SectorRaiz {
			get { return mSectorRaiz; }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Pertenece(Punto); }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Colisionan(Caja); }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Colisionan(Figura2D.Caja); }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Pertenece(Recta); }
		}

		public Quadtree(int Niveles, Caja2D Caja)
		{
			if (Niveles > 1) {
				mNiveles = Niveles;

				mSectorRaiz = new SectorQuadtree(Niveles, Caja);
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (NEW): Un octree debe tener al menos dos niveles:" + Constants.vbNewLine + "Niveles=" + Niveles);
			}
		}

		public void Refrescar()
		{
			mSectorRaiz.Refrescar();
		}

		public SectorQuadtree Sector(Punto2D Punto)
		{
			return Sector(this, Punto);
		}

		public SectorQuadtree Sector(Recta2D Recta)
		{
			return Sector(this, Recta);
		}

		public SectorQuadtree Sector(Caja2D Caja)
		{
			return Sector(this, Caja);
		}

		public SectorQuadtree Sector(Figura2D Figura2D)
		{
			return Sector(this, Figura2D);
		}

		public SectorQuadtree[] Sectores(Recta2D Recta)
		{
			return Sectores(this, Recta);
		}

		public static SectorQuadtree Sector(Quadtree Quadtree, Punto2D Punto)
		{
			int Resultado = Quadtree.SectorRaiz.Pertenece(ref Punto);
			SectorQuadtree S = Quadtree.SectorRaiz;

			if (Resultado != -2) {
				if (Resultado == -1) {
					return S;
				} else {
					while (Resultado != -2 && Resultado != -1) {
						Resultado = S.Pertenece(ref Punto);
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.SubSectores[Resultado];
							}
						} else {
							return S.Padre;
						}
					}

					return S;
				}
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (PERTENECE): El punto especificado no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Punto=" + Punto.ToString() + Constants.vbNewLine + "Espacio=" + Quadtree.Espacio.ToString());
			}
		}

		public static SectorQuadtree Sector(Quadtree Quadtree, Recta2D Recta)
		{
			int Resultado = 0;
			SectorQuadtree S = Quadtree.SectorRaiz;

			Resultado = Quadtree.SectorRaiz.Pertenece(ref Recta);

			if (Resultado != -2) {
				if (Resultado == -1) {
					return S;
				} else {
					while (Resultado != -2 && Resultado != -1) {
						Resultado = S.Pertenece(ref Recta);
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.SubSectores[Resultado];
							}
						} else {
							return S.Padre;
						}
					}

					return S;
				}
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (PERTENECE): La recta especificada no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Punto=" + Recta.ToString() + Constants.vbNewLine + "Espacio=" + Quadtree.Espacio.ToString());
			}
		}

		public static SectorQuadtree[] Sectores(Quadtree Quadtree, Recta2D Recta)
		{
			SectorQuadtree S = Quadtree.SectorRaiz;
			List<SectorQuadtree> Retorno = new List<SectorQuadtree>();

			if (Quadtree.SectorRaiz.Espacio.Pertenece(Recta)) {
				InterseccionRecta(Recta, S, ref Retorno);

				return Retorno.ToArray();
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (PERTENECE): La recta especificado no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Recta=" + Recta.ToString() + Constants.vbNewLine + "Espacio=" + Quadtree.Espacio.ToString());
			}
		}

		public static void InterseccionRecta(Recta2D Recta, SectorQuadtree Sector, ref List<SectorQuadtree> ListaRetorno)
		{
			if (!Sector.EsHoja) {
				foreach (SectorQuadtree Hijo in Sector.SubSectores) {
					if (Hijo.Espacio.Pertenece(Recta)) {
						InterseccionRecta(Recta, Hijo, ref ListaRetorno);
					}
				}
			} else {
				if (Sector.Espacio.Pertenece(Recta))
					ListaRetorno.Add(Sector);
			}
		}

		public static SectorQuadtree Sector(Quadtree Quadtree, Caja2D Caja)
		{
			int Resultado = Quadtree.SectorRaiz.Pertenece(ref Caja);
			SectorQuadtree S = Quadtree.SectorRaiz;

			//PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
			if (Resultado != -2) {
				if (Resultado == -1) {
					return Quadtree.SectorRaiz;
				} else {
					do {
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.SubSectores[Resultado];
							}
						} else {
							return S.Padre;
						}
						Resultado = S.Pertenece(ref Caja);
					} while (Resultado != -2 && Resultado != -1);

					return S;
				}
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (PERTENECE): La caja especificada no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Caja=" + Caja.ToString() + Constants.vbNewLine + "Espacio=" + Quadtree.Espacio.ToString());
			}
		}

		public static SectorQuadtree Sector(Quadtree Quadtree, ref Figura2D Figura2D)
		{
			int Resultado = Quadtree.SectorRaiz.Pertenece(ref Figura2D);
			SectorQuadtree S = Quadtree.SectorRaiz;

			//PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
			if (Resultado != -2) {
				if (Resultado == -1) {
					return Quadtree.SectorRaiz;
				} else {
					do {
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.SubSectores[Resultado];
							}
						} else {
							return S.Padre;
						}
						Resultado = S.Pertenece(ref Figura2D);
					} while (Resultado != -2 && Resultado != -1);

					return S;
				}
			} else {
				throw new ExcepcionGeometrica2D("OCTREE (PERTENECE): La figura especificada no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Figura2D=" + Figura2D.ToString() + Constants.vbNewLine + "Espacio=" + Quadtree.Espacio.ToString());
			}
		}

		public override string ToString()
		{
			return "{Quadtree de espacio=" + mSectorRaiz.Espacio.ToString() + "}";
		}
	}
}


